### DISCUSSION TITLE
Display driver for TTGO LilyGo T7789 1.14 " Led Screen (see #576)

### DISCUSSION THREAD

**Discussion by louisvangeldrop** [Moddable-OpenSource/moddable#601] [General]
Since the last conversation I have tried to use the info from TFT-ESP project. 
The setup-file is: Setup25_TTGO_T_Display.h
And the init file: ST7789_Init.h

So I have updated the register-settings as follows:

 [
            "0x11, 0,",
            "0xff, 120,",
            "0x13, 0,",
            "0x36,1,0x08,",
            "0xb6,2,0x0a,0x82,",
            "0x3a,1,0x55,",
            "0xff,10,",
            "0xb2,5,0x0c,0x0c,0x00, 0x33,0x33,",
            "0xb7,1,0x35,",
            "0xbb,1,0x28,",
            "0xc0,1,0x0c,",
            "0xc2,2,0x01,0xFF,",
            "0xc3,1,0x10,",
            "0xc4,1,0x20,",
            "0xc6,1,0x0f,",
            "0xd0,2,0xa4,0xa1,",
            "0xe0,14,0xd0,0x00,0x02,0x07,0x0a,0x28,0x32,0x44,0x42,0x06,0x0e,0x12,0x14,0x17,",
            "0xe1,14,0xd0,0x00,0x02,0x07,0x0a,0x28,0x31,0x54,0x47,0x0e,0x1c,0x17,0x1b,0x1e,",
            "0x21,0,",
            "0x2a,4,0x00,0x00,0x00,0xE5,",
            "0x2b,4,0x00,0x00,0x01,0x3F,",
            "0xff,120,",
            "0x29,0,",
            "0xff,120,"
        ]

However it crashes immediately, when calling the config.Sreen({})

**Comment by phoddie** [comment]
Your message does not indicate what JSON element in the manifest the registers above are contained in. It could be:

```
"registers": [
```

Or 

```
			"registers_append": [
```

If it is `"registers"` then you need to include a terminator in the array after the delay. Without that, the `ili9341Init` function will blindly read memory past your data structure.

```
"0x29,0,",
"0xff,120,"
"0xff,0"
]
```
```

**Comment by louisvangeldrop** [comment]
"defines": {
        "spi": {
            "hz": 40000000,
            "mosi": 19,
            "sck": 18,
            "mode": 3
        },
"ili9341": {
            "spi_port": "HSPI_PORT",
            "width": 135,
            "height": 240,
            "cs_pin": 5,
            "dc_pin": 16,
            "rst_pin": 23,
            "backlight_pin": 4,
            "column_offset": 40,
            "row_offset": 53,
            "registers": [
"0x11, 0,",
"0xff, 120,",
"0x13, 0,",
"0x36,1,0x08,",
"0xb6,2,0x0a,0x82,",
"0x3a,1,0x55,",
"0xff,10,",
"0xb2,5,0x0c,0x0c,0x00, 0x33,0x33,",
"0xb7,1,0x35,",
"0xbb,1,0x28,",
"0xc0,1,0x0c,",
"0xc2,2,0x01,0xFF,",
"0xc3,1,0x10,",
"0xc4,1,0x20,",
"0xc6,1,0x0f,",
"0xd0,2,0xa4,0xa1,",
"0xe0,14,0xd0,0x00,0x02,0x07,0x0a,0x28,0x32,0x44,0x42,0x06,0x0e,0x12,0x14,0x17,",
"0xe1,14,0xd0,0x00,0x02,0x07,0x0a,0x28,0x31,0x54,0x47,0x0e,0x1c,0x17,0x1b,0x1e,",
"0x21,0,",
"0x2a,4,0x00,0x00,0x00,0xE5,",
"0x2b,4,0x00,0x00,0x01,0x3F,",
"0xff,120,",
"0x29,0,",
"0xff,120,"
]}

**Comment by phoddie** [comment]
Thanks. So you need to add the `"0xff,0"` as noted above.

**Comment by louisvangeldrop** [comment]
Clock example freeezes at line 38: render.end()

Manifest.json:

{
	"include": [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_commodetto.json",
		"$(MODDABLE)/examples/manifest_net.json",
		"$(MODULES)/drivers/st7789/manifest.json"
	],
	"modules": {
		"*": "./main"
	},
	"resources": {
		"*-alpha": "./digits"
	},
	"config": {
		"sntp": "pool.ntp.org",
		"screen": "st7789"
	},
	"defines": {
		"spi": {
			"hz": 40000000,
			"mosi": 19,
			"sck": 18,
			"mode": 3
		},
		"ili9341": {
			"spi_port": "HSPI_HOST",
			"width": 135,
			"height": 240,
			"cs_pin": 5,
			"dc_pin": 16,
			"rst_pin": 23,
			"backlight_pin": 4,
			"column_offset": 40,
			"row_offset": 53,
			"registers": []
		}
	}
}

**Comment by phoddie** [comment]
I suggested you **add** a line at the end of the `registers` element.  It seems you just **emptied out** the registers element instead. That seems very unlikely to work correctly. 

The $MODDABLE/examples/commodetto/clock example works as expected on Moddable Two and the simulator.

This does not appear to be an issue with the Moddable SDK but, instead, a request for help using the ILI9341 (MIPI) display driver with another display. Therefore, I'm moving this from Issues to Discussions.

**Comment by louisvangeldrop** [comment]
I have added 0xff,0 to the register-section. Still the examlpe freezes at the render.end() function.

**Reply by louisvangeldrop** [reply]
@phoddie . I thought that the ST7789-driver was supported in the Moddable SDK. It is the reason that I bought this device.

**Comment by louisvangeldrop** [comment]
I have changed the width to 240 and the height to 135, and now the render.end() doesn't freeze the ESP32. However the render.end() call swichtes the screen off.

**Comment by mkellner** [comment]
Looking at the registers closely, I see in `ST7789_Init.h` near the end:

```
  writecommand(ST7789_CASET);    // Column address set
  writedata(0x00);
  writedata(0x00);
  writedata(0x00);
  writedata(0xE5);    // 239

  writecommand(ST7789_RASET);    // Row address set
  writedata(0x00);
  writedata(0x00);
  writedata(0x01);
  writedata(0x3F);    // 319
```

This particular driver configuration is for a 240 x 320 screen. This was translated to the Moddable registers as:

```
"0x2a,4,0x00,0x00,0x00,0xE5,",
"0x2b,4,0x00,0x00,0x01,0x3F,",
```

Since this display is 240 x 135, we'll need to change these to something like

```
"0x2a,4,0x00,0x00,0x00,0xE5,",
"0x2b,4,0x00,0x00,0x00,0x87,",
```

Or depending on how the display was made, maybe it is 135 x 240:

```
"0x2a,4,0x00,0x00,0x00,0x87,",
"0x2b,4,0x00,0x00,0x00,0xE5,",
```

There may be other registers that need changing - the ST7789 is a driver chip, not a screen, so it can be used in a variety of ways. The Moddable SDK does support ST7789, but probably not in every use of that chip.

For example, this 240 x 135 IPS display may be the same as on the LilyGo?

![image](https://user-images.githubusercontent.com/6822704/110892378-ccfa3b80-82a8-11eb-8d49-453f2ad66d59.png)

It works with these registers:
```
            "registers": [
                "0x01, 0,",
                "0xff, 150,",
                "0x36, 1, 0x70,",
                "0x3A, 1, 0x05,",
                "0x21, 0,",
                "0x11, 0,",
                "0x29, 0,",
                "0xff, 100,",
                "0x2B, 4, 0x00, 0x35, 0x00, 0xbb,",
                "0x2A, 4, 0x00, 0x28, 0x01, 0x17,",
                "0xff, 100,"
            ]
```
See `0x2B` and `0x2A` lines? They are different from above. From pixel 54 (0x35) to 187 (0xbb) - and 40 (0x28) to 279 (0x117).

Is there sample code with the TTGO LilyGo T7789? A working set of example code would be the go-to source for which registers will work for that particular ST7789/display combination.

**Reply by louisvangeldrop** [reply]
@michael. I am using the samples from this website: https://github.com/Xinyuan-LilyGO/TTGO-T-Display/tree/master/TFT_eSPI with Setup25_TTGO_T_Display.h. Most samples do work.

Since the driver is the same as ili931, I switched to screen:ili9341 i.s.o. st7789. Reason is, that I don't know how the registers-settings are handled with the different manifest.jsnon files. One in the drivers-librarry and one in my own folder.

The backlight of the LED is on, but render.end() switches it off. In the Instrumentation window I can see that 30 frames/sec and 192k pixels are drawn. But the screen stays black.

**Reply by louisvangeldrop** [reply]
I have used your suggestions - also started to study this document: https://www.displayfuture.com/Display/datasheet/controller/ST7789V.pdf

The TTGO device I am using looks very much the same as the Pico Display. However the registers-settings do not help( black screen).
I will look at the latest ili9341-example and see if I can use that example for further testing.

**Reply by mkellner** [reply]
The black screen may be because the backlight isn't enabled or the sense of the backlight pin isn't right for your device. The following is the complete define for the Pico display. Things you will need to change: `backlight_pin`, `cs_pin` and `dc_pin` to match your device. The `spi_port` should be `VSPI_HOST`. Depending on how the backlight is wired, it may need `backlight_on` to be changed. Also, the speed here may be too high for your device. Maybe try `27000000` for the `hz`.

```
		"ili9341": {
			"hz": 64000000,
			"width": 240,
			"height": 135,
			"cs_pin": 17,
			"dc_pin": 16,
			"spi_port": 0,
			"backlight_pin": 20,
			"backlight_on": 1,
			"column_offset": 40,
			"row_offset": 53,
			"registers": [
				"0x01, 0,",
				"0xff, 150,",
				"0x36, 1, 0x70,",
				"0x3A, 1, 0x05,",
				"0x21, 0,",
				"0x11, 0,",
				"0x29, 0,",
				"0xff, 100,", 
				"0x2B, 4, 0x00, 0x35, 0x00, 0xbb,",
				"0x2A, 4, 0x00, 0x28, 0x01, 0x17,",
				"0xff, 100,"
			]
		}
```

**Reply by louisvangeldrop** [reply]
I am using the io/spi/ili9341/main.js example and adapted the init-settings to the st7789 chip. Still the TFT backlight is on, but nothing displayed.

**Reply by louisvangeldrop** [reply]
I am using this adapted code for the ST7789-chip: ( Still the LED remains black. The backlight is visible)

/*
* Copyright (c) 2021  Moddable Tech, Inc.
*
*   This file is part of the Moddable SDK.
*
*   This work is licensed under the
*       Creative Commons Attribution 4.0 International License.
*   To view a copy of this license, visit
*       <https://creativecommons.org/licenses/by/4.0>.
*   or send a letter to Creative Commons, PO Box 1866,
*   Mountain View, CA 94042, USA.
*
*/
// import Analog from "embedded:io/analog";
// import DigitalBank from "embedded:io/digitalbank";
// import I2C from "embedded:io/i2c";
// import PWM from "embedded:io/pwm";
// import Serial from "embedded:io/serial";

// see http://www.espruino.com/modules/ST7789.js for init-settings

import Digital from "embedded:io/digital";
import SPI from "embedded:io/spi";

import Timer from "timer";

const pins = { bl: 4 /*backlight*/, dc: 16 /*data*/, ce: 5/*chip select*/, rst: 23 /*reset*/ }

const backlight = new Digital({
	pin: pins.bl,
	mode: Digital.Output,
});
backlight.write(1);
// }

const dc = new Digital({
	pin: pins.dc,
	mode: Digital.Output,
});

const ce = new Digital({
	pin: pins.ce,
	mode: Digital.Output,
});

const rst = new Digital({
	pin: pins.rst,
	mode: Digital.Output,
});

const spi = new SPI({
	hz: 40_000_000,
	// in:-1,		//miso
	out: 19,	// mosi
	clock: 18,
	active: 0,
	mode: 3,
	io: SPI,
	port: 1
});

const LCD_WIDTH = 135;
const LCD_HEIGHT = 240;
const COLSTART = 52;
const ROWSTART = 40;
const ST7789_INIT_CODE = [  //  see http://www.espruino.com/modules/ST7789.js
	// CMD, D0,D1,D2...
	[0x11, 0],     	//SLPOUT (11h):
	[0xff, 120],	//wait 120 msec
	[0x13],
	// This is an unrotated screen
	[0x36, 0],     // MADCTL
	// These 2 rotate the screen by 180 degrees
	//0x36,0xC0],     // MADCTL
	//0x37,[0,80]],   // VSCSAD (37h): Vertical Scroll Start Address of RAM

	[0x3A, 0x55],  // COLMOD - interface pixel format - 16bpp
	[0xff, 10],
	[0xB2, [0xC, 0xC, 0, 0x33, 0x33]], // PORCTRL (B2h): Porch Setting
	[0xB7, 0],     // GCTRL (B7h): Gate Control
	[0xBB, 0x3E],  // VCOMS (BBh): VCOM Setting 
	[0xC2, 1],     // VDVVRHEN (C2h): VDV and VRH Command Enable
	[0xC3, 0x19],  // VRHS (C3h): VRH Set 
	[0xC4, 0x20],  // VDVS (C4h): VDV Set
	[0xC5, 0xF],   // VCMOFSET (C5h): VCOM Offset Set .
	[0xD0, [0xA4, 0xA1]],   // PWCTRL1 (D0h): Power Control 1 
	[0xe0, [0x70, 0x15, 0x20, 0x15, 0x10, 0x09, 0x48, 0x33, 0x53, 0x0B, 0x19, 0x15, 0x2a, 0x2f]],   // PVGAMCTRL (E0h): Positive Voltage Gamma Control
	[0xe1, [0x70, 0x15, 0x20, 0x15, 0x10, 0x09, 0x48, 0x33, 0x53, 0x0B, 0x19, 0x15, 0x2a, 0x2f]],   // NVGAMCTRL (E1h): Negative Voltage Gamma Contro
	[0x29], // DISPON (29h): Display On 
	[0x21] // INVON (21h): Display Inversion On
	// End
	[0, 0]// 255/*DATA_LEN = 255 => END*/
];

function init() {
	doCmd(0x01)			// software reset
	// rst.write(1)
	doCmd(0xff, 120)
	doCmd(0x11)			// exit sleep
	doCmd(0xff, 20)
	for (let cmd of ST7789_INIT_CODE) {
		doCmd(cmd[0], cmd[1])
	}
	doCmd(0xff, 20)
}

init()

fillRect(0xFFFF, 0, LCD_WIDTH, 0, LCD_HEIGHT);
fillRect(0xF800, 0, LCD_WIDTH, 0, LCD_HEIGHT);
fillRect(0x001F, 0, LCD_WIDTH, 0, LCD_HEIGHT);
fillRect(0x07E0, 0, LCD_WIDTH, 0, LCD_HEIGHT);
fillRect(0xF81F, 0, 120, 0, 160);
fillRect(0x07FF, 120, 240, 160, 320);

setPixel(0xff00ff, 50, 50)

function setPixel(c, x, y) {
	ce.write(0)
	doCmd(0x2A, (COLSTART + x) >> 8, COLSTART + x, (COLSTART + x) >> 8, COLSTART + x);
	doCmd(0x2B, (ROWSTART + y) >> 8, ROWSTART + y, (ROWSTART + y) >> 8, (ROWSTART + y));
	dc.write(1)
	writeCmd(0x2c, c >> 8, c);
	ce.write(1)
}

function fillRect(color, xMin, xMax, yMin, yMax) {
	ce.write(0)
	doCmd(0x2A, (COLSTART + xMin) >> 8, (COLSTART + xMin) & 0xff, (COLSTART + xMax) >> 8, (COLSTART + xMax) & 0xff);
	doCmd(0x2B, (ROWSTART + yMin) >> 8, (ROWSTART + yMin) & 0xff, (ROWSTART + yMax) >> 8, (ROWSTART + yMax) & 0xff);
	doCmd(0x2C);
	// doCmd({ data: String.fromCharCode(color >> 8, color), count: (xMax - xMin + 1) * (yMax - yMin + 1) });
	dc.write(1)
	const lines = 4;
	const pixels = new Uint16Array((xMax - xMin) * lines);
	pixels.fill((color >> 8) | (color << 8));		// big endian

	for (yMax -= yMin; yMax > 0; yMax -= lines)
		spi.write(pixels);
	doCmd(0x29)
	ce.write(1)
}

function fill(color, xMin, xMax, yMin, yMax) {
	doCmd(0x2a, xMin >> 8, xMin & 0xff, xMax >> 8, xMax & 0xff);
	doCmd(0x2b, yMin >> 8, yMin & 0xff, yMax >> 8, yMax & 0xff);
	doCmd(0x2c);

	dc.write(1);

	const lines = 4;
	const pixels = new Uint16Array((xMax - xMin) * lines);
	pixels.fill((color >> 8) | (color << 8));		// big endian

	for (yMax -= yMin; yMax > 0; yMax -= lines)
		spi.write(pixels);
}

function doCmd(register, ...data) {
	spi.flush(true);
	dc.write(0);
	if (register === 0xff) { Timer.delay(data) }
	else {
		spi.write(Uint8Array.of(register));
		if (data.length) {
			dc.write(1);
			spi.write(Uint8Array.from(data));
		}
	}
}

function writeCmd(register, ...data) {
	spi.write(Uint8Array.of(register));
	if (data.length) {
		dc.write(1);
		spi.write(Uint8Array.from(data));
	}
}

**Reply by louisvangeldrop** [reply]
@mkellner . Sorry I forgot to mention that I have tried the settings you suggested. Unfortunately no result.

**Comment by mkellner** [comment]
The TTGO display device is a little hard to come by. It arrived today.

Please try the following files. Put them in a folder in `$MODDABLE/build/devices/esp32/target/lilygo`.
Then build an app like `examples/piu/balls` or `examples/piu/cards` with `mcconfig -d -m -p esp32/lilygo`.


[Archive.zip](https://github.com/Moddable-OpenSource/moddable/files/6187773/Archive.zip)

This is from the `examples/piu/neon-light`.
![image](https://user-images.githubusercontent.com/6822704/112111453-71199780-8b71-11eb-92ba-19222d868869.png)

**Reply by louisvangeldrop** [reply]
SUPER!!!!. It works. Thank you very much for this solution. I am so unbelievable happy. I am still curious to know what change made the solution.

**Reply by mkellner** [reply]
We tried a lot in this thread.
The speed at 40Mhz was too high. I reduced it to 27MHz (other speeds may work).
The sense of `backlight_on` needed to be set. Some devices use `0` to turn on the backlight.
spi `mode` should be `0` (ie. left to the default).

`VSPI_HOST` was used for the `spi_port`. It corresponds to:
```
pin: 5 - CS
pin: 18 - SCK
pin: 19 - MISO
pin: 23 - MOSI
```
